core/trino-main/src/main/java/io/trino/type/TypeCoercion.java

```diff
@@ -85,11 +85,32 @@ public boolean isTypeOnlyCoercion(Type source, Type result)
         if (source instanceof DecimalType && result instanceof DecimalType) {
             DecimalType sourceDecimal = (DecimalType) source;
             DecimalType resultDecimal = (DecimalType) result;
-            boolean sameDecimalSubtype = (sourceDecimal.isShort() && resultDecimal.isShort())
-                    || (!sourceDecimal.isShort() && !resultDecimal.isShort());
+            boolean sameDecimalSubtype = sourceDecimal.isShort() == resultDecimal.isShort();
             boolean sameScale = sourceDecimal.getScale() == resultDecimal.getScale();
-            boolean sourcePrecisionIsLessOrEqualToResultPrecision = sourceDecimal.getPrecision() <= resultDecimal.getPrecision();
-            return sameDecimalSubtype && sameScale && sourcePrecisionIsLessOrEqualToResultPrecision;
+            boolean noPrecisionLoss = sourceDecimal.getPrecision() <= resultDecimal.getPrecision();
+            return sameDecimalSubtype && sameScale && noPrecisionLoss;
+        }
+
+        if (source instanceof TimestampWithTimeZoneType sourceTimestampWithTimeZone && result instanceof TimestampWithTimeZoneType targetTimestampWithTimeZone) {
```

Comment:
sourceTimestampWithTimeZone -> sourceType
targetTimestampWithTimeZone -> targetType

These long variables names do not add to readability